sw/device/tests/dma_abort.c

```diff
@@ -0,0 +1,157 @@
+  dif_pinmux_pad_attr_t in_attr = {
+      .slew_rate = 0,
+      .drive_strength = 0,
+      .flags = kDifPinmuxPadAttrPullResistorEnable |
+               kDifPinmuxPadAttrPullResistorUp};
+  for (uint32_t i = 0; i <= ARRAYSIZE(spi_host0_direct_pads); ++i) {
+    CHECK_DIF_OK(dif_pinmux_pad_write_attrs(&pinmux, spi_host0_direct_pads[i],
+                                            kDifPinmuxPadKindDio, in_attr,
+                                            &out_attr));
+  }
+}
+
+bool test_main(void) {
+  // Initialize the pinmux.
+  CHECK_DIF_OK(dif_pinmux_init(
+      mmio_region_from_addr(TOP_DARJEELING_PINMUX_AON_BASE_ADDR), &pinmux));
+  pinmux_testutils_init(&pinmux);
+
+  // Initialise DMA.
+  CHECK_DIF_OK(
+      dif_dma_init(mmio_region_from_addr(TOP_DARJEELING_DMA_BASE_ADDR), &dma));
+
+  // Setup pinmux if required, enable weak pull-up on relevant pads
+  setup_pads_spi_host0();  // direct
+
+  // Setup spi host configuration
+  CHECK_DIF_OK(dif_spi_host_init(
+      mmio_region_from_addr(TOP_DARJEELING_SPI_HOST0_BASE_ADDR), &spi_host));
+  init_spi_host(&spi_host, (uint32_t)kClockFreqHiSpeedPeripheralHz);
+
+  // DV sync message
+  LOG_INFO("spi host configuration complete");
+
+  dif_spi_host_segment_t host_operations[1] = {
+      {.type = kDifSpiHostSegmentTypeRx,
+       .tx = {.width = kDifSpiHostWidthStandard,
+              .buf = &received_data[TX_SIZE],
+              .length = TX_SIZE}},
+  };
+
+  // Dummy assignment to avoid any unused variable warnings
+  kShaDigestExpData[0] = 0;
+
+  // Issue the SPI transaction
+  CHECK_DIF_OK(dif_spi_host_start_transaction(
+      &spi_host, /*csid=*/0, host_operations, ARRAYSIZE(host_operations)));
+
+  // Configure the DMA to read from SPI in hardware-handshake mode
+  dif_dma_transaction_t transaction_1 = {
+      .source = {.address = TOP_DARJEELING_SPI_HOST0_BASE_ADDR +
+                            SPI_HOST_RXDATA_REG_OFFSET,
+                 .asid = kDifDmaOpentitanInternalBus},
+      .destination = {.address = (uint32_t)&received_data[0],
+                      .asid = kDifDmaOpentitanInternalBus},
+      .total_size = TX_SIZE,
+      .chunk_size = CHUNK_SIZE,
+      .width = kDifDmaTransWidth4Bytes};
+
+  dif_dma_handshake_t handshake_config = {.fifo_auto_increment = false,
+                                          .memory_auto_increment = true,
+                                          .direction_from_mem_to_fifo = false};
+
+  CHECK_DIF_OK(dif_dma_memory_range_set(&dma, TOP_DARJEELING_RAM_MAIN_BASE_ADDR,
+                                        TOP_DARJEELING_RAM_MAIN_SIZE_BYTES));
+
+  CHECK_DIF_OK(dif_dma_handshake_irq_enable(&dma, 0x2));
+  CHECK_DIF_OK(dif_dma_configure(&dma, transaction_1));
+  CHECK_DIF_OK(dif_dma_handshake_enable(&dma, handshake_config));
+
+  CHECK_DIF_OK(dif_dma_start(&dma, kDifDmaSha256Opcode));
+
+  // Wait a few cycles to let the DMA run
+  for (int i = 0; i < 200; ++i) {
+    __asm__("nop");
+  }
+
+  CHECK_DIF_OK(dif_dma_abort(&dma));
```

Comment:
Timing the abort relative to the start is quite tricky: as it turns out, the current delay with a `nop` loop of 200 iterations is not sufficient to for having a DMA transfer (LSIO-triggered by `spi_host`) before the abort (at least when I ran it with Xcelium and a seed of `1`). Extending the loop to 2500 iterations means the abort comes *after* the DMA transfer, when the DMA is idle again.

The most interesting (and probably critical) case, I think, is aborting during a DMA transfer. To hit this case, I suggest letting the vseq decide the instant (which could be randomized to happen before, during, or after the DMA transfer) and signal that to SW (e.g., with a backdoor symbol write or with an interrupt).

Reply:
This has been resolved:
> For the abort testcase, the hardware vseq waits until the first TLUL transfer and adds then a random delay that is in between the duration of the testcase. If that delay is passed, it releases the software barrier to let the software trigger the abort. After the aborted testcase, the software sets up a new transfer and reads data from the SPI (via the DMA).